### PR TITLE
Add ExasolPool (allow more than 1 connection)

### LINKED ISSUE #28: Add a connection pool
It seems like this driver only supports 1 connection in a connection pool. I can see there are some tests and code related to multiple connections in a pool. but when initializing the driver, it uses a hardcoded value of 1 when it creates the pool.

Is there any particular reason we can't use a large pool size?

### PR DESCRIPTION
**Pull request by pj-spoelders**
Closes #28 
Adds an `ExasolPool` class that functions as a `Connection Pool`

### CODE REVIEW

File: integration-test/testcases/pool.basic.spec.ts [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+      await runQueryXNumberOfTimesAndCheckResult(amountOfRequests, poolToQuery, schemaName);
+
+      await poolToQuery.drain();
+      await poolToQuery.clear();
+
+      await setupClient.close();
+    });
+
+    afterAll(async () => {
+      //  await container.stop();
```
Comment (kaklakariada) [inline integration-test/testcases/pool.basic.spec.ts:240]:
```suggestion
```

File: integration-test/testcases/pool.basic.spec.ts [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+      await setupClient.connect();
+
+      await setupClient.execute('CREATE SCHEMA ' + schemaName);
+      await setupClient.execute('CREATE TABLE ' + schemaName + '.TEST_TABLE(x INT)');
+      await setupClient.execute('INSERT INTO ' + schemaName + '.TEST_TABLE VALUES (15)');
+
+      const amountOfRequests = 20;
+
+      await runQueryXNumberOfTimesAndCheckResult(amountOfRequests, poolToQuery, schemaName);
```
Comment (kaklakariada) [inline integration-test/testcases/pool.basic.spec.ts:231]: Test name says 100, but actually only runs 20 requests.

File: integration-test/testcases/pool.basic.spec.ts [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+      const poolToQuery = new ExasolPool(factory, {
+        host: container.getHost(),
+        port: container.getMappedPort(8563),
+        user: 'sys',
+        password: 'exasol',
+        encryption: false,
+        min: 1,
+        max: 10,
+      });
```
Comment (kaklakariada) [inline integration-test/testcases/pool.basic.spec.ts:96]: Code duplication: extract this into a method and pass min and max as arguments. This avoids duplicating code.

File: src/lib/sql-client.ts [hunk 4]
```diff
@@ -225,7 +225,13 @@ export class ExasolDriver implements IExasolDriver {
...
     getCancel?: CetCancelFunction | undefined,
-    responseType?: 'default' | 'raw'
+    responseType?: 'default' | 'raw',
+  ): Promise<QueryResult | SQLResponse<SQLQueriesResponse>>;
+  async query(
+    sqlStatement: string,
+    attributes?: Partial<Attributes> | undefined,
+    getCancel?: CetCancelFunction | undefined,
+    responseType?: 'default' | 'raw' | undefined,
   ): Promise<QueryResult | SQLResponse<SQLQueriesResponse>> {
```
Comment (kaklakariada) [inline src/lib/sql-client.ts:235]: What is the reason for this change? I only see that `responseType` can now be `undefined`. But `?` already marks the parameter as optional, so `| undefined` should not be necessary.

File: src/lib/sql-pool.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    getCancel?: CetCancelFunction | undefined,
+    responseType?: 'default' | 'raw',
+  ): Promise<QueryResult | SQLResponse<SQLQueriesResponse>> {
+    {
+      let exasolClient;
+      try {
+        exasolClient = await this.internalPool.acquire();
+
+        const response = await exasolClient.query(sqlStatement, attributes, getCancel, responseType);
+        await this.internalPool.release(exasolClient);
```
Comment (kaklakariada) [inline src/lib/sql-pool.ts:86]: can you do the release also in  a `finally` block?

File: src/lib/sql-pool.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+import { Attributes } from './commands';
+import { CetCancelFunction } from './sql-client.interface';
+import { SQLQueriesResponse, SQLResponse } from './types';
+//https://www.npmjs.com/package/generic-pool
+//https://www.npmjs.com/package/@types/generic-pool
+export interface PoolConfig {
+  min: number;
+  max: number;
+}
+export function getPool(websocketFactory: websocketFactory, config: Partial<Config> & Partial<PoolConfig>, logger: ILogger) {
```
Comment (kaklakariada) [inline src/lib/sql-pool.ts:14]:
```suggestion
function getPool(websocketFactory: websocketFactory, config: Partial<Config> & Partial<PoolConfig>, logger: ILogger) {
```
Does this need to be exported?

File: src/lib/sql-pool.ts [hunk 7]
```diff
@@ -0,0 +1,103 @@
+import { createPool, Factory, Pool } from 'generic-pool';
+import { Config, ExasolDriver, websocketFactory } from './sql-client';
+import { ILogger, LogLevel, Logger } from './logger/logger';
+import { QueryResult } from './query-result';
+import { Attributes } from './commands';
+import { CetCancelFunction } from './sql-client.interface';
+import { SQLQueriesResponse, SQLResponse } from './types';
+//https://www.npmjs.com/package/generic-pool
+//https://www.npmjs.com/package/@types/generic-pool
```
Comment (kaklakariada) [inline src/lib/sql-pool.ts:9]:
```suggestion
```

File: src/lib/sql-pool.ts [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+export interface PoolConfig {
+  min: number;
+  max: number;
+}
+export function getPool(websocketFactory: websocketFactory, config: Partial<Config> & Partial<PoolConfig>, logger: ILogger) {
+  // export interface Factory<T> {
+  //   create(): Promise<T>;
+  //   destroy(client: T): Promise<void>;
+  //   validate?(client: T): Promise<boolean>;
+  // }
```
Comment (kaklakariada) [inline src/lib/sql-pool.ts:19]:
```suggestion
```

File: src/lib/sql-pool.ts [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+  async function createClient() {
+    const exasolClient: ExasolDriver = new ExasolDriver(websocketFactory, config, logger);
+    await exasolClient.connect();
+    return Promise.resolve(exasolClient);
```
Comment (kaklakariada) [inline src/lib/sql-pool.ts:23]:
```suggestion
    return await exasolClient.connect();
```
Should do the same, not sure ;)

File: src/lib/sql-pool.ts [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    },
+  };
+  const poolOpts: PoolConfig = {
+    max: config.max ?? 5, // maximum size of the pool
+    min: config.min ?? 0, // minimum size of the pool
+  };
+  const tempPool = createPool(poolFactory, poolOpts);
+  return tempPool;
+}
+export class ExasolPool {
```
Comment (kaklakariada) [inline src/lib/sql-pool.ts:45]: Add JSDoc comments to class, constructor and methods

File: src/lib/sql-pool.ts [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+  private internalPool: Pool<ExasolDriver>;
+  private logger: ILogger;
+  constructor(
+    websocketFactory: websocketFactory,
+    config: Partial<Config> & Partial<PoolConfig>,
+    logger: ILogger = new Logger(LogLevel.Debug),
+  ) {
+    this.logger = logger;
+    this.internalPool = getPool(websocketFactory, config, logger);
+  }
```
Comment (kaklakariada) [inline src/lib/sql-pool.ts:55]: The type definition of the config parameter is a bit complicated. It is possible to do it like this, but maybe just adding a second parameter for the pool configuration would be better.
Happy to discuss!

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: integration-test/node/pool.runner.spec.ts [hunk 12]
```diff
@@ -0,0 +1,6 @@
+import { WebSocket } from 'ws';
+import { ExaWebsocket } from '../../src/lib/connection';
+import { basicPoolTests } from '../testcases/pool.basic.spec';
+basicPoolTests('Browser', (url) => {
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply integration-test/node/pool.runner.spec.ts:4]: Good catch!

File: doc/user_guide/user_guide.md [hunk 13]
```diff
@@ -138,3 +138,97 @@ console.log(queryResult.getRows()[0]['X']);
...
+
+NodeJs:
+
+```js
+import { ExaWebsocket, ExasolPool } from "@exasol/exasol-driver-ts";
+import { WebSocket } from 'ws';
+
+const pool = new ExasolPool((url) => {
+  return new WebSocket(url) as ExaWebsocket;
+}, {
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply doc/user_guide/user_guide.md:172]: Feel free to create the ticket!

File: src/lib/sql-pool.ts (hunk 5 above)
Comment (pj-spoelders) [reply src/lib/sql-pool.ts:79]: Yes.

File: src/lib/sql-pool.ts (hunk 9 above)
Comment (pj-spoelders) [reply src/lib/sql-pool.ts:16]: Yep, it does, I don't think it's as readable but it does. I'll change it.

File: integration-test/testcases/pool.basic.spec.ts (hunk 2 above)
Comment (pj-spoelders) [reply integration-test/testcases/pool.basic.spec.ts:225]: Fixed. Good catch.

File: integration-test/testcases/pool.basic.spec.ts [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+
+export const basicPoolTests = (name: string, factory: websocketFactory) =>
+  describe(name, () => {
+    const randomId = new RandomUuid();
+    let container: StartedTestContainer;
+    jest.setTimeout(7000000);
+    let schemaName = '';
+
+    beforeAll(async () => {
+      container = await new GenericContainer('exasol/docker-db:7.1.22')
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply integration-test/testcases/pool.basic.spec.ts:15]: Done

File: src/lib/sql-pool.ts (hunk 5 above)
Comment (pj-spoelders) [reply src/lib/sql-pool.ts:79]: Changed it.

File: src/lib/sql-client.ts (hunk 4 above)
Comment (pj-spoelders) [reply src/lib/sql-client.ts:235]: It's a typescript specific thing. I'll call you and show you. We can also discuss the separate config for the pool properties then.

File: integration-test/testcases/pool.basic.spec.ts [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+        encryption: false,
+      });
+
+      await setupClient.connect();
+
+      await setupClient.execute('CREATE SCHEMA ' + schemaName);
+      await setupClient.execute('CREATE TABLE ' + schemaName + '.TEST_TABLE(x INT)');
+      await setupClient.execute('INSERT INTO ' + schemaName + '.TEST_TABLE VALUES (15)');
+
+      const data = await poolToQuery.query('SELECT x FROM ' + schemaName + '.TEST_TABLE');
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply integration-test/testcases/pool.basic.spec.ts:65]: Ok .. I'll keep it in mind but I'll leave the the tests as is now. Think this is fine.

File: integration-test/testcases/pool.basic.spec.ts (hunk 3 above)
Comment (pj-spoelders) [reply integration-test/testcases/pool.basic.spec.ts:93]: Done.

File: src/lib/sql-pool.ts (hunk 10 above)
Comment (pj-spoelders) [reply src/lib/sql-pool.ts:45]: Done

File: src/lib/sql-pool.ts (hunk 11 above)
Comment (pj-spoelders) [reply src/lib/sql-pool.ts]: Made the property names more clear. As discussed ..

File: doc/user_guide/user_guide.md [hunk 16]
```diff
@@ -230,5 +230,5 @@ await pool.clear();
...
 | `schema`           |       string       |                     | Exasol schema name.                                                                                                                     |
-| `min`              |       number       |          0          | Minimum amount of active connections.                                                                                                   |
-| `max`              |       number       |          5          | Maximum amount of active connections.                                                                                                   |
+| `minimumPoolSize`  |       number       |          0          | Minimum amount of active connections.                                                                                                   |
+| `maximumPoolSize`  |       number       |          5          | maximumPoolSizeimum amount of active connections.                                                                                       |
```
Comment (kaklakariada) [inline doc/user_guide/user_guide.md:234]:
```suggestion
| `minimumPoolSize`  |       number       |          0          | Minimum number of active connections.                                                                                                   |
| `maximumPoolSize`  |       number       |          5          | Maximum number of active connections.                                                                                       |
```

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: doc/user_guide/user_guide.md (hunk 16 above)
Comment (pj-spoelders) [reply doc/user_guide/user_guide.md]: Fixed it, good catch ..

**Review by kaklakariada — APPROVED** [review]
